BaragonAgentService/src/main/java/com/hubspot/baragon/agent/lbs/FilesystemConfigHelper.java

```diff
@@ -53,57 +54,59 @@ public void remove(String serviceId, boolean reloadConfigs) throws LbAdapterExec
-    final String serviceId = context.getService().getServiceId();
-    final String oldServiceId;
-    if (replaceServiceId.isPresent()) {
-      oldServiceId = replaceServiceId.get();
+  public void apply(ServiceContext context, Optional<BaragonService> maybeOldService, boolean revertOnFailure) throws InvalidConfigException, LbAdapterExecuteException, IOException, MissingTemplateException {
+    final BaragonService service = context.getService();
+    final BaragonService oldService;
+    if (maybeOldService.isPresent()) {
+      oldService = maybeOldService.get();
     } else {
-      oldServiceId = serviceId;
+      oldService = service;
     }
 
-    LOG.info(String.format("Going to apply %s: %s", serviceId, Joiner.on(", ").join(context.getUpstreams())));
-    final boolean oldServiceExists = configsExist(oldServiceId);
-    final boolean previousConfigsExist = configsExist(serviceId);
+    LOG.info(String.format("Going to apply %s: %s", service.getServiceId(), Joiner.on(", ").join(context.getUpstreams())));
+    final boolean oldServiceExists = configsExist(oldService);
+    final boolean previousConfigsExist = configsExist(service);
 
-    if (configGenerator.generateConfigsForProject(context).equals(readConfigs(oldServiceId))) {
+    Collection<BaragonConfigFile> newConfigs = configGenerator.generateConfigsForProject(context);
+
+    if (newConfigs.equals(readConfigs(oldService))) {
       LOG.info("    Configs are unchanged, skipping apply");
       return;
     }
 
     // Backup configs
     if (revertOnFailure) {
-      backupConfigs(serviceId);
+      backupConfigs(service);
       if (oldServiceExists) {
-        backupConfigs(oldServiceId);
+        backupConfigs(oldService);
       }
     }
 
     // Write & check the configs
     try {
       if (context.isPresent()) {
-        writeConfigs(configGenerator.generateConfigsForProject(context));
+        writeConfigs(newConfigs);
         //If the new service id for this base path is different, remove the configs for the old service id
-        if (oldServiceExists && !oldServiceId.equals(serviceId)) {
-          remove(oldServiceId, false);
+        if (oldServiceExists && !oldService.getServiceId().equals(service.getServiceId())) {
+          remove(oldService, false);
         }
       } else {
-        remove(serviceId, false);
+        remove(service, false);
       }
 
       adapter.checkConfigs();
     } catch (Exception e) {
-      LOG.error("Caught exception while writing configs for " + serviceId + ", reverting to backups!", e);
+      LOG.error("Caught exception while writing configs for " + service.getServiceId() + ", reverting to backups!", e);
 
       // Restore configs
       if (revertOnFailure) {
         if (oldServiceExists) {
-          restoreConfigs(oldServiceId);
+          restoreConfigs(oldService);
         }
         if (previousConfigsExist) {
-          restoreConfigs(serviceId);
+          restoreConfigs(service);
```

Comment:
I don't think there would be any side effects since we would just be copying from the same file, but I can add an extra check for oldService != service just to be sure